### PR TITLE
[Crypto Add] add secp256K1 Sign

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

The issue is that the default C# dotnet does not support SECP256K1 curve on macos platform. We added support to SECP256K1 signature `verify` method for macos but did not add its `sign` method. 

I found this issue when i was trying to run unit test here https://github.com/neo-project/neo-devpack-dotnet/blob/53a383ab5670d3c22d9761a113a50f7ef1a75ee0/tests/Neo.SmartContract.Framework.UnitTests/Services/CryptoTest.cs#L53

Fixes # (issue)

```
Test method Neo.SmartContract.Framework.UnitTests.Services.CryptoTest.Test_VerifySignatureWithMessage threw exception: 
System.PlatformNotSupportedException: The specified curve '1.3.132.0.10' or its parameters are not valid for this platform.
    at System.Security.Cryptography.EccSecurityTransforms.ImportParameters(ECParameters parameters)
   at System.Security.Cryptography.ECDsaImplementation.ECDsaSecurityTransforms.ImportParameters(ECParameters parameters)
   at System.Security.Cryptography.ECDsa.Create(ECParameters parameters)
   at Neo.SmartContract.Framework.UnitTests.Services.CryptoTest.Test_VerifySignatureWithMessage() in /Users/jinghuiliao/git/neo-devpack-dotnet/tests/Neo.SmartContract.Framework.UnitTests/Services/CryptoTest.cs:line 70
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)

```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] TestSecp256k1

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jim8y** [comment]
Ok, macos, windows, ubuntu all have different behaviors,,,,,

**Comment by Jim8y** [comment]
@shargon good to go now

**Comment by vang1ong7ang** [comment]
btw how to ensure the `IsOSX` can be well tested?

File: src/Neo/Cryptography/Crypto.cs [hunk 1]
```diff
@@ -48,20 +50,41 @@ public static byte[] Hash256(ReadOnlySpan<byte> message)
...
         /// </summary>
         /// <param name="message">The message to be signed.</param>
-        /// <param name="prikey">The private key to be used.</param>
-        /// <param name="pubkey">The public key to be used.</param>
+        /// <param name="priKey">The private key to be used.</param>
+        /// <param name="pubKey">The public key to be used (uncommpressed, 65 bytes).</param>
+        /// <param name="ecCurve">The <see cref="ECC.ECCurve"/> curve of the signature, default is <see cref="ECC.ECCurve.Secp256k1"/>.</param>
         /// <returns>The ECDSA signature for the specified message.</returns>
-        public static byte[] Sign(byte[] message, byte[] prikey, byte[] pubkey)
+        public static byte[] Sign(byte[] message, byte[] priKey, byte[] pubKey, ECC.ECCurve ecCurve = null)
```
[the remark this answers is not in the document]
Comment (vang1ong7ang) [reply src/Neo/Cryptography/Crypto.cs:57]: lets fix it in future prs

**Review by vang1ong7ang — APPROVED** [review]

File: src/Neo/Cryptography/Crypto.cs [hunk 2]
```diff
@@ -48,20 +50,40 @@ public static byte[] Hash256(ReadOnlySpan<byte> message)
...
+                var sBytes = signature[1].ToByteArrayUnsigned();
+
+                // Copy r and s into their respective parts of the signatureBytes array, aligning them to the right.
+                Buffer.BlockCopy(rBytes, 0, signatureBytes, 32 - rBytes.Length, rBytes.Length);
+                Buffer.BlockCopy(sBytes, 0, signatureBytes, 64 - sBytes.Length, sBytes.Length);
+                return signatureBytes;
+            }
+
+            var curve =
+                 ecCurve == null || ecCurve == ECC.ECCurve.Secp256r1 ? ECCurve.NamedCurves.nistP256 :
```
Comment (shargon) [inline src/Neo/Cryptography/Crypto.cs:79]:
```suggestion
                ecCurve == null || ecCurve == ECC.ECCurve.Secp256r1 ? ECCurve.NamedCurves.nistP256 :
```

**Review by shargon — DISMISSED** [review]

File: src/Neo/Cryptography/Crypto.cs [hunk 3]
```diff
@@ -48,20 +50,40 @@ public static byte[] Hash256(ReadOnlySpan<byte> message)
...
+
+                // Copy r and s into their respective parts of the signatureBytes array, aligning them to the right.
+                Buffer.BlockCopy(rBytes, 0, signatureBytes, 32 - rBytes.Length, rBytes.Length);
+                Buffer.BlockCopy(sBytes, 0, signatureBytes, 64 - sBytes.Length, sBytes.Length);
+                return signatureBytes;
+            }
+
+            var curve =
+                ecCurve == null || ecCurve == ECC.ECCurve.Secp256r1 ? ECCurve.NamedCurves.nistP256 :
+                ecCurve == ECC.ECCurve.Secp256k1 ? ECCurve.CreateFromFriendlyName("secP256k1") :
```
Comment (AnnaShaleva) [inline src/Neo/Cryptography/Crypto.cs:80]: How about making the `ECCurve.CreateFromFriendlyName("secP256k1")` a `public static` property to allocate this object once, and then reuse it everywhere?

File: src/Neo/Cryptography/Crypto.cs [hunk 4]
```diff
@@ -48,20 +50,40 @@ public static byte[] Hash256(ReadOnlySpan<byte> message)
         /// Signs the specified message using the ECDSA algorithm.
         /// </summary>
         /// <param name="message">The message to be signed.</param>
-        /// <param name="prikey">The private key to be used.</param>
-        /// <param name="pubkey">The public key to be used.</param>
+        /// <param name="priKey">The private key to be used.</param>
+        /// <param name="ecCurve">The <see cref="ECC.ECCurve"/> curve of the signature, default is <see cref="ECC.ECCurve.Secp256k1"/>.</param>
```
Comment (AnnaShaleva) [inline src/Neo/Cryptography/Crypto.cs:54]: Not quite right, the default curve is secp256r1, not a Koblitz curve.

File: src/Neo/Cryptography/Crypto.cs (hunk 3 above)
Comment (Jim8y) [reply src/Neo/Cryptography/Crypto.cs:80]: maybe a cache system in another pr

**Review by shargon — APPROVED** [review]

**Pull request merged**